In [41]:
import pandas as pd
iris1 = pd.read_csv('iris1.csv')
iris2 = pd.read_csv('iris2.csv')

In [42]:
iris1.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           NaN          3.0           NaN          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

In [43]:
iris2.head()

sepal.length  sepal.width  petal.length  petal.width     variety
0           5.7          2.8           4.1          1.3  Versicolor
1           6.3          3.3           6.0          2.5   Virginica
2           5.8          2.7           5.1          1.9   Virginica
3           7.1          3.0           5.9          2.1   Virginica
4           6.3          2.9           5.6          1.8   Virginica

In [44]:
# NaN means not a number, null value

# iris1.dropna()      # Remove all rows with NULL values from the Dataframe
# iris1.fillna(0)     # Replace NULL values with the number 0

# Not only use for whole dataset, but also use for signle column
iris1['sepal.length'].fillna(iris1['sepal.length'].mean(), inplace = True)
iris1['petal.length'].fillna(iris1['petal.length'].mean(), inplace = True)
iris1

sepal.length  sepal.width  petal.length  petal.width     variety
0        5.10000          3.5      1.400000          0.2      Setosa
1        5.47449          3.0      2.863265          0.2      Setosa
2        4.70000          3.2      1.300000          0.2      Setosa
3        4.60000          3.1      1.500000          0.2      Setosa
4        5.00000          3.6      1.400000          0.2      Setosa
..           ...          ...           ...          ...         ...
94       5.60000          2.7      4.200000          1.3  Versicolor
95       5.70000          3.0      4.200000          1.2  Versicolor
96       5.70000          2.9      4.200000          1.3  Versicolor
97       6.20000          2.9      4.300000          1.3  Versicolor
98       5.10000          2.5      3.000000          1.1  Versicolor

[99 rows x 5 columns]

In [45]:
df = pd.concat([iris1, iris2], axis=0)      # Row wise concatenation
# df = pd.concat([iris1, iris2], axis=1)      # Column wise concatenation 
df

sepal.length  sepal.width  petal.length  petal.width    variety
0        5.10000          3.5      1.400000          0.2     Setosa
1        5.47449          3.0      2.863265          0.2     Setosa
2        4.70000          3.2      1.300000          0.2     Setosa
3        4.60000          3.1      1.500000          0.2     Setosa
4        5.00000          3.6      1.400000          0.2     Setosa
..           ...          ...           ...          ...        ...
46       6.70000          3.0      5.200000          2.3  Virginica
47       6.30000          2.5      5.000000          1.9  Virginica
48       6.50000          3.0      5.200000          2.0  Virginica
49       6.20000          3.4      5.400000          2.3  Virginica
50       5.90000          3.0      5.100000          1.8  Virginica

[150 rows x 5 columns]

In [46]:
df['variety'].unique()

array(['Setosa', 'Versicolor', 'Virginica'], dtype=object)

In [47]:
df[df['variety'] == 'Setosa']

sepal.length  sepal.width  petal.length  petal.width variety
0        5.10000          3.5      1.400000          0.2  Setosa
1        5.47449          3.0      2.863265          0.2  Setosa
2        4.70000          3.2      1.300000          0.2  Setosa
3        4.60000          3.1      1.500000          0.2  Setosa
4        5.00000          3.6      1.400000          0.2  Setosa
5        5.40000          3.9      1.700000          0.4  Setosa
6        4.60000          3.4      1.400000          0.3  Setosa
7        5.00000          3.4      1.500000          0.2  Setosa
8        4.40000          2.9      1.400000          0.2  Setosa
9        4.90000          3.1      1.500000          0.1  Setosa
10       5.40000          3.7      1.500000          0.2  Setosa
11       4.80000          3.4      1.600000          0.2  Setosa
12       4.80000          3.0      1.400000          0.1  Setosa
13       4.30000          3.0      1.100000          0.1  Setosa
14       5.80000          4.0      1.200000          0.2  Setosa
15       5.70000          4.4      1.500000          0.4  Setosa
16       5.40000          3.9      1.300000          0.4  Setosa
17       5.10000          3.5      1.400000          0.3  Setosa
18       5.70000          3.8      1.700000          0.3  Setosa
19       5.10000          3.8      1.500000          0.3  Setosa
20       5.40000          3.4      1.700000          0.2  Setosa
21       5.10000          3.7      1.500000          0.4  Setosa
22       4.60000          3.6      1.000000          0.2  Setosa
23       5.10000          3.3      1.700000          0.5  Setosa
24       4.80000          3.4      1.900000          0.2  Setosa
25       5.00000          3.0      1.600000          0.2  Setosa
26       5.00000          3.4      1.600000          0.4  Setosa
27       5.20000          3.5      1.500000          0.2  Setosa
28       5.20000          3.4      1.400000          0.2  Setosa
29       4.70000          3.2      1.600000          0.2  Setosa
30       4.80000          3.1      1.600000          0.2  Setosa
31       5.40000          3.4      1.500000          0.4  Setosa
32       5.20000          4.1      1.500000          0.1  Setosa
33       5.50000          4.2      1.400000          0.2  Setosa
34       4.90000          3.1      1.500000          0.2  Setosa
35       5.00000          3.2      1.200000          0.2  Setosa
36       5.50000          3.5      1.300000          0.2  Setosa
37       4.90000          3.6      1.400000          0.1  Setosa
38       4.40000          3.0      1.300000          0.2  Setosa
39       5.10000          3.4      1.500000          0.2  Setosa
40       5.00000          3.5      1.300000          0.3  Setosa
41       4.50000          2.3      1.300000          0.3  Setosa
42       4.40000          3.2      1.300000          0.2  Setosa
43       5.00000          3.5      1.600000          0.6  Setosa
44       5.10000          3.8      1.900000          0.4  Setosa
45       4.80000          3.0      1.400000          0.3  Setosa
46       5.10000          3.8      1.600000          0.2  Setosa
47       4.60000          3.2      1.400000          0.2  Setosa
48       5.30000          3.7      1.500000          0.2  Setosa
49       5.00000          3.3      1.400000          0.2  Setosa

In [48]:
from sklearn.preprocessing import LabelEncoder

# Encode target labels with value between 0 and n_classes-1
label_encoder = LabelEncoder()
label_encoder

LabelEncoder()

In [49]:
df['variety'] = label_encoder.fit_transform(df['variety'])      # Fit label encoder and return encoded labels
df['variety']

0     0
1     0
2     0
3     0
4     0
     ..
46    2
47    2
48    2
49    2
50    2
Name: variety, Length: 150, dtype: int32

In [50]:
df.head()

sepal.length  sepal.width  petal.length  petal.width  variety
0       5.10000          3.5      1.400000          0.2        0
1       5.47449          3.0      2.863265          0.2        0
2       4.70000          3.2      1.300000          0.2        0
3       4.60000          3.1      1.500000          0.2        0
4       5.00000          3.6      1.400000          0.2        0

In [51]:
# df['vareity'].map({'Setosa' : 1, "Virginica" : 2, 'Versicolor' : 3})

In [52]:
# Seperating features from labels
X = df.drop(columns='variety')      # (Column) X = df.drop(columns='variety', axis = 1)
# X = df.drop([2, 3, 4], axis=0)      # (Row) X = df.drop([2, 3, 4], axis = 0)
X

sepal.length  sepal.width  petal.length  petal.width
0        5.10000          3.5      1.400000          0.2
1        5.47449          3.0      2.863265          0.2
2        4.70000          3.2      1.300000          0.2
3        4.60000          3.1      1.500000          0.2
4        5.00000          3.6      1.400000          0.2
..           ...          ...           ...          ...
46       6.70000          3.0      5.200000          2.3
47       6.30000          2.5      5.000000          1.9
48       6.50000          3.0      5.200000          2.0
49       6.20000          3.4      5.400000          2.3
50       5.90000          3.0      5.100000          1.8

[150 rows x 4 columns]

In [53]:
y = df['variety']
y

0     0
1     0
2     0
3     0
4     0
     ..
46    2
47    2
48    2
49    2
50    2
Name: variety, Length: 150, dtype: int32

In [54]:
# Spliting the dataset into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=20)
X_train

sepal.length  sepal.width  petal.length  petal.width
2            4.7          3.2           1.3          0.2
5            5.4          3.9           1.7          0.4
49           5.0          3.3           1.4          0.2
14           5.7          2.5           5.0          2.0
44           5.1          3.8           1.9          0.4
..           ...          ...           ...          ...
22           4.6          3.6           1.0          0.2
75           6.6          3.0           4.4          1.4
38           6.4          3.1           5.5          1.8
15           5.7          4.4           1.5          0.4
0            5.7          2.8           4.1          1.3

[105 rows x 4 columns]

In [55]:
# Create GaussianNB Object
from sklearn.naive_bayes import GaussianNB
nb_model = GaussianNB()
nb_model

GaussianNB()

In [56]:
# Train the classifier with the training sets
nb_model.fit(X_train, y_train)      # Fit Gaussian Naive Bayes according to X, y

GaussianNB()

In [57]:
# Calulate the mean accuracy
nb_model.score(X_train, y_train)

0.9809523809523809

In [58]:
nb_model.score(X_test, y_test)

0.8888888888888888

### Result

- The model is overfitting for test_size = 0.3, random_state = 20 (accuracy score for the train set 0.980) > (accuracy score for test set 0.889)

- (The Best Result) The model is overfitting for test_size = 0.4, random_state = 3 (accuracy score for the train set 0.944) > (accuracy score for test set 0.983)

- The model is overfitting for test_size = 0.2, random_state = 10 (accuracy score for the train set 0.950) > (accuracy score for test set 0.966)

In [59]:
# Other methods
nb_model.predict_proba(X_test)      # Return probability estimates for the test X

array([[1.00000000e+000, 3.42055131e-019, 6.43935593e-031],
       [3.80311088e-091, 9.99881735e-001, 1.18265421e-004],
       [4.86026489e-079, 9.99787720e-001, 2.12280006e-004],
       [1.49168867e-170, 3.18939716e-004, 9.99681060e-001],
       [2.17913891e-059, 9.99999789e-001, 2.11304446e-007],
       [4.76928357e-066, 9.99998007e-001, 1.99325159e-006],
       [1.59441263e-219, 1.40610948e-009, 9.99999999e-001],
       [1.00000000e+000, 4.68530865e-017, 2.74962696e-027],
       [1.77386761e-179, 5.07745797e-004, 9.99492254e-001],
       [1.00000000e+000, 3.06984098e-019, 1.49941201e-031],
       [3.74599730e-157, 2.81743703e-004, 9.99718256e-001],
       [6.59747442e-106, 9.72403283e-001, 2.75967165e-002],
       [8.45107069e-118, 9.95592031e-001, 4.40796895e-003],
       [1.00000000e+000, 1.92581034e-019, 4.60567184e-031],
       [1.00000000e+000, 6.83675826e-019, 3.39268410e-030],
       [2.70063560e-257, 4.98913994e-011, 1.00000000e+000],
       [1.00000000e+000, 7.19918025e-021

In [60]:
nb_model.predict(X_test)        # Perform classification on an array of test sets

array([0, 1, 1, 2, 1, 1, 2, 0, 2, 0, 2, 1, 1, 0, 0, 2, 0, 1, 2, 1, 1, 2,
       2, 0, 1, 1, 1, 0, 2, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 2, 1, 1, 0, 1,
       1])